MaxText/layers/linears.py

```diff
@@ -548,16 +551,15 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
 
     if self.config.capacity_factor > 0:
       # token dropping if needed
-      dispatch_mask, combine_mask = self.generate_masks(top_k_indices, softmax_probs)
+      weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)
```

Comment:
Why do you need this?

Reply:
`self.reshape_and_update_weights` will update weights from shape `(batch_size, seq_len, num_experts_per_tok)` to `(batch_size, seq_len, num_experts)`.  And this `update_weights = update_weights.at[index_update].set(weights)` will heavily impact the performance.

Reply:
+1 on performance impact. I just ran benchmark with this change and see perf slightly worse than the else block at CF=2. I believe this was mainly added for correctness.

Reply:
It's a quite big drop in training (please see [here](https://screenshot.googleplex.com/8cUszXRtgCHPbvz)). I am trying to understand why `softmax_probs *= combined_expert_mask` does not work as expected. As reshape_and_update_weights() did is to update the weights based on topk. And `combined_expert_mask` comes from top_k_indices. We have a unit test (here) for dropping, and it seems we haven't caught this bug. Taking a look.

Reply:
If we dont apply this line we get garbage outputs. 
{'rouge1': 0.0447, 'rouge2': 0.0, 'rougeL': 0.0447, 'rougeLsum': 0.0447, 'gen_len': 114987, 'gen_num': 100}

and all outputs being almost same. Which looks like correct weights are not selected. And yes the CL is for correctness and we still have to look into preformance

Reply:
I pulled the head, and add needed changes as discussed [here](https://github.com/AI-Hypercomputer/maxtext/pull/new/fix_weights). Without changing the weights, I was able to get

```
Successful requests: 100
Benchmark duration: 701.665692 s
Total input tokens: 21249
Total generated tokens: 29693
Request throughput: 0.14 requests/s
Input token throughput: 30.28 tokens/s
Output token throughput: 42.32 tokens/s
Mean TTFT: 439501.60 ms
Median TTFT: 460449.96 ms
P99 TTFT: 575143.83 ms
Mean TPOT: 4333.02 ms
Median TPOT: 3540.90 ms
P99 TPOT: 18755.05 ms
[nltk_data] Downloading package punkt to /home/ranran/nltk_data...
[nltk_data]   Package punkt is already up-to-date!

Results

{'rouge1': 44.5787, 'rouge2': 24.1149, 'rougeL': 31.4474, 'rougeLsum': 41.9571, 'gen_len': 110198, 'gen_num': 100}
```

Cmd:

```
python MaxText/maxengine_server.py MaxText/configs/base.yml load_parameters_path=gs://singhvijaya-maxtext-logs/8-22B-correct  max_prefill_predict_length=1024 max_target_length=2048 model_name=mixtral-8x22b ici_fsdp_parallelism=1 ici_autoregressive_parallelism=1 ici_tensor_parallelism=-1 scan_layers=false per_device_batch_size=20 attention=dot_product megablox=False quantization=int8 checkpoint_is_quantized=True quantize_kvcache=True capacity_factor=2


python JetStream/benchmarks/benchmark_serving.py   --tokenizer ~/maxtext/assets/tokenizer.mistral-v3 --save-result   --save-request-outputs   --request-outputs-file-path outputs.json   --num-prompts 100   --max-output-length 1024   --dataset openorca --run-eval True
```

Comparing the topk_weights & softmax, and I could see same values, and paste a few examples:

```
...
topk_weights: [[0 0 0 0 0 0.0966797 0.625 0]]
softmax_probs: [[0 0 0 0 0 0.0966797 0.625 0]]
topk_weights: [[0 0 0 0.291016 0 0 0 0.253906]]
softmax_probs: [[0 0 0 0.291016 0 0 0 0.253906]]
topk_weights: [[0 0 0.804688 0 0 0.0371094 0 0]]
softmax_probs: [[0 0 0.804688 0 0 0.0371094 0 0]]
topk_weights: [[0 0 0 0 0.172852 0 0.384766 0]]
softmax_probs: [[0 0 0 0 0.172852 0 0.384766 0]]
topk_weights: [[0 0 0 0 0 0.578125 0.152344 0]]
softmax_probs: [[0 0 0 0 0 0.578125 0.152344 0]]
topk_weights: [[0 0 0 0 0.202148 0.349609 0 0]]
softmax_probs: [[0 0 0 0 0.202148 0.349609 0 0]]
topk_weights: [[0 0 0 0.11084 0 0 0.617188 0]]
softmax_probs: [[0 0 0 0.11084 0 0 0.617188 0]]
topk_weights: [[0.417969 0 0 0 0.539062 0 0 0]]
softmax_probs: [[0.417969 0 0 0 0.539062 0 0 0]]
```

Reply:
I replicated Ran's setup on v6e-8 and still get 0 rouge on 20 samples
{'rouge1': 0.0, 'rouge2': 0.0, 'rougeL': 0.0, 'rougeLsum': 0.0, 'gen_len': 81347, 'gen_num': 20}

Reply:
What's the difference between two weights? Could you pull that branch, and run again (instead of update one line)? I am also curious what's the delta here. With that debug info, we would know the difference between two weights in a single run (if cannot reproduce).

Reply:
We filed a bug to compiler team.

Reply:
Removed this as this is not needed. The compiler team is investigating the behaviour.